tests/osmosis_highway_tunnel_bridge.osm

```diff
@@ -0,0 +1,215 @@
+  </node>
+  <node id='4' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86643916161' lon='5.8837517583' />
+  <node id='5' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86666971101' lon='5.88367034847' />
+  <node id='6' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.8665084999' lon='5.88367596294' />
+  <node id='7' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86660036776' lon='5.88367276348'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='8' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86659863952' lon='5.88354963736' />
+  <node id='9' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86634902586' lon='5.88407757914'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='10' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86653796864' lon='5.88406616865' />
+  <node id='11' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86642009356' lon='5.88393422859' />
+  <node id='12' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86642876086' lon='5.88424863894' />
+  <node id='13' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86642391683' lon='5.88407305638'>
+    <tag k='note' v='Bad node class 4/5' />
+  </node>
+  <node id='14' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86628228247' lon='5.88408160984' />
+  <node id='15' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86649376466' lon='5.88398476023' />
+  <node id='16' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.8664980983' lon='5.88422758608' />
+  <node id='17' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86649526357' lon='5.88406874766'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='18' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86611327022' lon='5.88532661728'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='19' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86666797756' lon='5.88531819557'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='20' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86600752836' lon='5.88532661728'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='21' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86677718476' lon='5.88521432787' />
+  <node id='22' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86660901291' lon='5.88531909079'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='23' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86664717616' lon='5.88547259351' />
+  <node id='24' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86657266235' lon='5.88531964268'>
+    <tag k='note' v='No match class 4/5' />
+  </node>
+  <node id='25' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86657090427' lon='5.88519467722' />
+  <node id='26' timestamp='2014-03-31T22:00:00Z' version='1' lat='51.86651016597' lon='5.88532059152'>
+    <tag k='note' v='Bad node class 4/5' />
```

Comment:
Imo this being an error would result in too many false positives.

Reply:
Can you explain why? It matches your original request as far as I know (see the comment about split ways)?

Also I didn't see too many false positives (tested on 12 extracts, [results](https://github.com/user-attachments/files/16559967/bridgetunnel.zip)), most of them are very close to the bridge heads where it's debatable where exactly the bridge starts, but this isn't limited to end or non-end nodes. Same as that this isn't related to having 1 or 2 non-bridge ways connected to a node because often it involves paths leading off of the bridge which is frequently symmetrical.

More than happy to exclude all end nodes if you can explain why, but the vast majority of valid cases are also like this so it may make the analyser pretty useless :)

Reply:
My idea was that it would only count as an error if like node 39 it had multiple end connections where both ways have the same highway tag. 

If this makes it too complicated I think just checking middle nodes would be good enough.

Reply:
The bt_ways only using middle nodes `nodes.id != bt_ways.nodes[1] AND
        nodes.id != bt_ways.nodes[array_length(bt_ways.nodes, 1)]` would not detect the specific case I mentioned anyway since the bridge way was split and not connected via a middle node.

Reply:
It would mostly make it more computationally intensive (always have to run it in full mode). And if a single end node would cause false positives, I have to consider that bridges are often symmetric, so then I would indeed also exclude double end nodes and only use middle nodes.

However, I'm primarily trying to understand why you expect the current method would give  too many false positives? Maybe there's a specific thing you have in mind, which may have other solutions (e.g. if it's footways, we can also just exclude those). Can you maybe give examples (see the results file shared earlier)? Or draw a hypothetical one?

Except at bridge heads, I cannot think of a way how a road can connect, via an end node, somewhere in the middle of the bridge, except if the bridge is interrupted (but then the bridge should be split in [bridge][regular][bridge]) or if the connecting way is at least partially a bridge too, even if just for a few meters until it reaches the existing bridge. Maybe very exotic constructions?

More than happy to make this change, but I'd like to understand first why it's needed :)

Reply:
They are not so much false positives but they're usually [incorrect bridge tags](https://www.openstreetmap.org/node/583110103) or [incorrect connections near the end of a bridge](https://www.openstreetmap.org/node/6336921071)  [(2)](https://www.openstreetmap.org/node/7891591583) which are not navigation breaking like a bridge being accidentally glued to another way. 

I don't mind if these are included if they're put in a different class or severity.

Reply:
Ok, now I understand why you requested it. Done. 
I went for the end node vs not end-node distinction because this allows us to run the analyser in diff mode, which is much more efficient.